llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/Passes/MaterializeEncoding.cpp

```diff
@@ -341,7 +422,10 @@ void populateMaterializeEncodingPatterns(
     MaterializeEncodingTypeConverter &typeConverter) {
 
   // Add all patterns for converting from encoded type to the materialized type
-  patterns.insert<MaterializeTiledMatmul, SetEncodingOpToPackOpConversion,
+  patterns.insert<MaterializeDPSOperation<linalg::FillOp>,
+                  MaterializeDPSOperation<linalg::MatmulOp>,
+                  MaterializeOperation<tensor::EmptyOp>,
```

Comment:
I think it's better to use benefits feature instead of the order. So maybe something like

```cpp
  const PatternBenefit kBaseBenefit = 1;
  patterns.insert<MaterializeOperation<tensor::EmptyOp>>(..., kBaseBenefit);

  // Try these specialized patterns first before resorting to the generic one.
  patterns.insert<MaterializeDPSOperation<linalg::FillOp>,
                  MaterializeDPSOperation<linalg::MatmulOp>,
  ...
  >(typeConverter, patterns.getContext(), kBaseBenefit.getBenefit() + 1);
```

Reply:
Oh, the ordering is just alphabetical. This textual ordering has no bearing on order of pattern application AFAIK.

Reply:
I was thinking that there are `MaterializeDPSOperation` and `MaterializeOperation`, and if we'll add a pattern like `MaterializeOperation<linalg::LinalgOp>` to the list. Now I just found that it would be interface case, so it does not matter anymore...

Reply:
I am not sure we can write an interface pattern for linalg ops anyway

Reply:
Yes, we can do it using `OpInterfaceRewritePattern`. LinalgOp is an interface, we can write an interface pattern if needed. (or for DestinationStyleOpInterface) but it looks like we'll have different logic for fill and matmul, then it's not needed.

Here is an example that reorder CastOpInterface ops and vector broadcast ops: https://github.com/llvm/llvm-project/blob/c60a8d9af38d65accc3c06dd5a4975f89f8d2724/mlir/lib/Dialect/Vector/Transforms/VectorTransforms.cpp#L1283-L1285

Reply:
:) I meant to say that I can think of writing an interface pattern that can lower any linalg op with operands having encoding to a sequence of one or more ops whose operands dont have encoding